main-command/src/main/scala/sbt/internal/client/BspClient.scala

```diff
@@ -18,24 +18,39 @@ import scala.sys.process.Process
 import scala.util.control.NonFatal
 
 class BspClient private (sbtServer: Socket) {
+  private var terminated = false
 
   private def transferTo(input: InputStream, output: OutputStream): Unit = {
     val buffer = Array.ofDim[Byte](1024)
     while (true) {
       val size = input.read(buffer)
-      output.write(buffer, 0, size)
-      output.flush()
+      if (size == -1) synchronized {
+        output.close()
+        terminated = true
+        notify()
+      } else {
+        output.write(buffer, 0, size)
+        output.flush()
+      }
     }
   }
 
   private def run(): Exit = {
     try {
-      val redirection = new Thread {
+      val socketToStd = new Thread {
         override def run(): Unit = transferTo(sbtServer.getInputStream, System.out)
       }
 
-      redirection.start()
-      transferTo(System.in, sbtServer.getOutputStream)
+      val stdToSocket = new Thread {
+        override def run(): Unit = transferTo(System.in, sbtServer.getOutputStream)
+      }
+
+      socketToStd.start()
```

Comment:
I think socketToStd and stdToSocket both look like the should be daemon threads. It maybe doesn't matter but is good practice.

Reply:
Thank you @eatkins for this review. It's the first time I hear about daemon process.